lib/balena.js

```diff
@@ -0,0 +1,350 @@
+		if (opts.dataDirectory == null) {
+			opts.dataDirectory = settings.get('dataDirectory');
+		}
+	}
+
+	const auth = new BalenaAuth(opts);
+	const request = getRequest(Object.assign({}, opts, { auth }));
+	const pine = getPine(Object.assign({}, opts, { auth, request }));
+	const pubsub = new PubSub();
+
+	const sdk = {};
+	const deps = {
+		settings,
+		request,
+		auth,
+		pine,
+		pubsub,
+		sdkInstance: sdk,
+	};
+
+	Object.keys(sdkTemplate).forEach(function(moduleName) {
+		Object.defineProperty(sdk, moduleName, {
+			enumerable: true,
+			configurable: true,
+			get() {
+				const moduleFactory = sdkTemplate[moduleName]();
+				// We need the delete first as the current property is read-only
+				// and the delete removes that restriction
+				delete this[moduleName];
+				return (this[moduleName] = moduleFactory(deps, opts));
+			},
+		});
+	});
+
+	/**
+	 * @typedef Interceptor
+	 * @type {object}
+	 * @memberof balena.interceptors
+	 *
+	 * @description
+	 * An interceptor implements some set of the four interception hook callbacks.
+	 * To continue processing, each function should return a value or a promise that
+	 * successfully resolves to a value.
+	 *
+	 * To halt processing, each function should throw an error or return a promise that
+	 * rejects with an error.
+	 *
+	 * @property {function} [request] - Callback invoked before requests are made. Called with
+	 * the request options, should return (or resolve to) new request options, or throw/reject.
+	 *
+	 * @property {function} [response] - Callback invoked before responses are returned. Called with
+	 * the response, should return (or resolve to) a new response, or throw/reject.
+	 *
+	 * @property {function} [requestError] - Callback invoked if an error happens before a request.
+	 * Called with the error itself, caused by a preceeding request interceptor rejecting/throwing
+	 * an error for the request, or a failing in preflight token validation. Should return (or resolve
+	 * to) new request options, or throw/reject.
+	 *
+	 * @property {function} [responseError] - Callback invoked if an error happens in the response.
+	 * Called with the error itself, caused by a preceeding response interceptor rejecting/throwing
+	 * an error for the request, a network error, or an error response from the server. Should return
+	 * (or resolve to) a new response, or throw/reject.
+	 */
+
+	/**
+	 * @summary Array of interceptors
+	 * @member {Interceptor[]} interceptors
+	 * @public
+	 * @memberof balena
+	 *
+	 * @description
+	 * The current array of interceptors to use. Interceptors intercept requests made
+	 * internally and are executed in the order they appear in this array for requests,
+	 * and in the reverse order for responses.
+	 *
+	 * @example
+	 * balena.interceptors.push({
+	 * 	responseError: function (error) {
+	 * 		console.log(error);
+	 * 		throw error;
+	 * 	})
+	 * });
+	 */
+	Object.defineProperty(sdk, 'interceptors', {
+		/** @returns Interceptor[] */
```

Comment:
Even though it might not feel 100% correct, we might need to add `@private` here to not cause changes in the docs.